### ISSUE TITLE
warning: manually reimplementing `div_ceil`, only pops-up when the macro is used in a application but not in the crate sourcecode.

### ISSUE THREAD

**Issue by vDorst** [rust-lang/rust#136995]
I am using https://github.com/embassy-rs/atomic-pool in my code which has a macro `pool!()`.
When used in a application like below and run `cargo +nightly clippy`.

This output is given.
```
    Checking typenum v1.17.0
    Checking stable_deref_trait v1.2.0
    Checking portable-atomic v1.10.0
    Checking as-slice v0.2.1
    Checking generic-array v0.14.7
    Checking generic-array v0.13.3
    Checking generic-array v0.12.4
    Checking as-slice v0.1.5
    Checking atomic-pool v2.0.0
    Checking bug2 v0.1.0 (<snip>/bug2)
warning: manually reimplementing `div_ceil`
 --> src/main.rs:3:1
  |
3 | pool!(VarName: [u32; 4]);
  | ^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#manual_div_ceil
  = note: `#[warn(clippy::manual_div_ceil)]` on by default
  = note: this warning originates in the macro `pool` (in Nightly builds, run with -Z macro-backtrace for more info)

warning: manually reimplementing `div_ceil`
 --> src/main.rs:3:1
  |
3 | pool!(VarName: [u32; 4]);
  | ^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#manual_div_ceil
  = note: this warning originates in the macro `pool` (in Nightly builds, run with -Z macro-backtrace for more info)

warning: `bug2` (bin "bug2") generated 2 warnings
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.58s
```

I tried this code:

```rust
use atomic_pool::pool;

pool!(VarName: [u32; 4]);
fn main() {
    println!("Hello, world!");
}

Cargo.toml
[package]
name = "bug2"
version = "0.1.0"
edition = "2021"

[dependencies]
atomic-pool = "2.0.0"
```

But when I clone the `atomic-pool` repo and then run ``cargo +nightly clippy` in that repo.
I was expecting to see the same warning to pop-up.
But it does not.
Is this right?

```
# cargo +nightly clippy
    Checking atomic-pool v2.0.0 (<snip>/atomic-pool)
warning: a `const` item should not be interior mutable
  --> src/atomic_bitset.rs:15:9
   |
15 |         const Z: AtomicU32 = AtomicU32::new(0);
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = help: consider making this a static item
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#declare_interior_mutable_const
   = note: `#[warn(clippy::declare_interior_mutable_const)]` on by default

warning: you should consider adding a `Default` implementation for `PoolStorageImpl<T, N, K>`
  --> src/lib.rs:40:5
   |
40 | /     pub const fn new() -> Self {
41 | |         Self {
42 | |             used: AtomicBitset::new(),
43 | |             data: [Self::UNINIT; N],
44 | |         }
45 | |     }
   | |_____^
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#new_without_default
   = note: `#[warn(clippy::new_without_default)]` on by default
help: try adding this
   |
34 + impl<T, const N: usize, const K: usize> Default for PoolStorageImpl<T, N, K>
35 + where
36 +     [AtomicU32; K]: Sized,
37 +  {
38 +     fn default() -> Self {
39 +         Self::new()
40 +     }
41 + }
   |

warning: a `const` item should not be interior mutable
  --> src/lib.rs:38:5
   |
38 |     const UNINIT: UnsafeCell<MaybeUninit<T>> = UnsafeCell::new(MaybeUninit::uninit());
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#declare_interior_mutable_const

warning: this `match` expression can be replaced with `?`
  --> src/lib.rs:86:17
   |
86 |           let p = match P::get().alloc() {
   |  _________________^
87 | |             Some(p) => p,
88 | |             None => return None,
89 | |         };
   | |_________^ help: try instead: `P::get().alloc()?`
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#question_mark
   = note: `#[warn(clippy::question_mark)]` on by default

warning: unsafe function's docs are missing a `# Safety` section
   --> src/lib.rs:100:5
    |
100 |     pub unsafe fn from_raw(ptr: NonNull<P::Item>) -> Self {
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#missing_safety_doc
    = note: `#[warn(clippy::missing_safety_doc)]` on by default

warning: `atomic-pool` (lib) generated 5 warnings (run `cargo clippy --fix --lib -p atomic-pool` to apply 2 suggestions)
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.12s
```

### Meta

`rustc --version --verbose`:
```
rustc 1.86.0-nightly (a9730c3b5 2025-02-05)
binary: rustc
commit-hash: a9730c3b5f84a001c052c60c97ed0765e9ceac04
commit-date: 2025-02-05
host: x86_64-unknown-linux-gnu
release: 1.86.0-nightly
LLVM version: 19.1.7
```

**Comment by hanna-kruppe** [comment]
Most lints work post macro expansion, and the macro is not used in the library itself so the lint can’t fire there. Except in the library’s tests, so `cargo +nightly clippy -- --test` should emit the lint.

**Comment by workingjubilee** [comment]
I don't foresee a universe in which we inspect macros that aren't invoked in order to lint on their hypothetical expansions.
